File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/GamepadEx.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.hardware.Gamepad;

/**
 * A wrapper around a Gamepad. Provides additional utility methods, such as didFall and didRise to detect falling-/rising-edge button presses.
 */
public class GamepadEx {
	/**
	 * A gamepad button.
	 */
	public enum Button {
		A,
		B,
		X,
		Y,
		LB,
		RB,
		UP,
		DOWN,
		LEFT,
		RIGHT,
		BACK,
		GUIDE,
		START,
		LS,
		RS,
	}

	private final Gamepad previous = new Gamepad();
	private final Gamepad current = new Gamepad();

	/**
	 * Update the gamepad state. Should be called every loop, before using anything that uses the gamepad.
	 *
	 * @param gamepad Gamepad to update state from.
	 */
	public void tick(Gamepad gamepad) {
		previous.copy(current);
		current.copy(gamepad);
	}

	/**
	 * Checks for a rising-edge press from a button.
	 *
	 * @param button Button to check.
	 * @return Returns true if the button did rise (was pressed).
	 */
	public boolean didRise(Button button) {
		switch (button) {
			case A:
				return !previous.a && current.a;
			case B:
				return !previous.b && current.b;
			case X:
				return !previous.x && current.x;
			case Y:
				return !previous.y && current.y;
			case LB:
				return !previous.left_bumper && current.left_bumper;
			case RB:
				return !previous.right_bumper && current.right_bumper;
			case UP:
				return !previous.dpad_up && current.dpad_up;
			case DOWN:
				return !previous.dpad_down && current.dpad_down;
			case LEFT:
				return !previous.dpad_left && current.dpad_left;
			case RIGHT:
				return !previous.dpad_right && current.dpad_right;
			case BACK:
				return !previous.back && current.back;
			case GUIDE:
				return !previous.guide && current.guide;
			case START:
				return !previous.start && current.start;
			case LS:
				return !previous.left_stick_button && current.left_stick_button;
			case RS:
				return !previous.right_stick_button && current.right_stick_button;
			default:
				return false;
		}
	}

	/**
	 * Checks for a falling-edge press from a button.
	 *
	 * @param button Button to check.
	 * @return Returns true if the button did fall (was released).
	 */
	public boolean didFall(Button button) {
		switch (button) {
			case A:
				return previous.a && !current.a;
			case B:
				return previous.b && !current.b;
			case X:
				return previous.x && !current.x;
			case Y:
				return previous.y && !current.y;
			case LB:
				return previous.left_bumper && !current.left_bumper;
			case RB:
				return previous.right_bumper && !current.right_bumper;
			case UP:
				return previous.dpad_up && !current.dpad_up;
			case DOWN:
				return previous.dpad_down && !current.dpad_down;
			case LEFT:
				return previous.dpad_left && !current.dpad_left;
			case RIGHT:
				return previous.dpad_right && !current.dpad_right;
			case BACK:
				return previous.back && !current.back;
			case GUIDE:
				return previous.guide && !current.guide;
			case START:
				return previous.start && !current.start;
			case LS:
				return previous.left_stick_button && !current.left_stick_button;
			case RS:
				return previous.right_stick_button && !current.right_stick_button;
			default:
				return false;
		}
	}

	public boolean isPressed(Button button) {
		switch (button) {
			case A:
				return current.a;
			case B:
				return current.b;
			case X:
				return current.x;
			case Y:
				return current.y;
			case LB:
				return current.left_bumper;
			case RB:
				return current.right_bumper;
			case UP:
				return current.dpad_up;
			case DOWN:
				return current.dpad_down;
			case LEFT:
				return current.dpad_left;
			case RIGHT:
				return current.dpad_right;
			case BACK:
				return current.back;
			case GUIDE:
				return current.guide;
			case START:
				return current.start;
			case LS:
				return current.left_stick_button;
			case RS:
				return current.right_stick_button;
			default:
				return false;
		}
	}

	/**
	 * Get the left stick's X value.
	 *
	 * @return Returns the left stick's X value.
	 */
	public double getLeftX() {
		return current.left_stick_x;
	}

	/**
	 * Get the left stick's Y value.
	 *
	 * @return Returns the left stick's Y value.
	 */
	public double getLeftY() {
		return current.left_stick_y;
	}

	/**
	 * Get the right stick's X value.
	 *
	 * @return Returns the right stick's X value.
	 */
	public double getRightX() {
		return current.right_stick_x;
	}

	/**
	 * Get the right stick's Y value.
	 *
	 * @return Returns the right stick's Y value.
	 */
	public double getRightY() {
		return current.right_stick_y;
	}
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/GamepadEx.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.hardware.Gamepad

private const val TRIGGER_THRESHOLD = 0.2

typealias Event = Pair<GamepadEx.Button, GamepadEx.RisingOrFalling>
typealias Handler = () -> Unit

class GamepadEx(private val gamepad: Gamepad) {
	enum class Button {
		A,
		B,
		X,
		Y,
		LB,
		RB,
		LT,
		RT,
		UP,
		DOWN,
		LEFT,
		RIGHT,
		BACK,
		GUIDE,
		START,
		LS,
		RS,
	}

	enum class RisingOrFalling {
		RISING,
		FALLING,
	}

	private var previous = Gamepad()
	private var current = Gamepad()

	private val handlers: MutableMap<Event, MutableSet<Handler>> = mutableMapOf()

	val leftX: Double get() = current.left_stick_x.toDouble()
	val leftY: Double get() = current.left_stick_y.toDouble()
	val rightX: Double get() = current.right_stick_x.toDouble()
	val rightY: Double get() = current.right_stick_y.toDouble()
	val LT: Double get() = current.left_trigger.toDouble()
	val RT: Double get() = current.right_trigger.toDouble()

	fun tick() {
		previous.copy(current)
		current.copy(gamepad)

		for (button in Button.values()) {
			if (didRise(button))
				handlers[button to RisingOrFalling.RISING]?.forEach { it() }
			else if (didFall(button))
				handlers[button to RisingOrFalling.FALLING]?.forEach { it() }
		}
	}

	fun addButtonHandler(event: Event, handler: Handler) {
		val list = handlers[event] ?: run {
			handlers[event] = mutableSetOf()
			handlers[event]!!
		}
		list += handler
	}

	fun removeButtonHandler(event: Event, handler: Handler) {
		val list = handlers[event] ?: return
		list.remove(handler)
	}

	fun getButton(button: Button): Boolean = getButtonPair(button).second
	private fun didRise(button: Button): Boolean =
		getButtonPair(button).let { !it.first && it.second }

	private fun didFall(button: Button): Boolean =
		getButtonPair(button).let { it.first && !it.second }

	private fun getButtonPair(button: Button) =
		when (button) {
			Button.A -> previous.a to current.a
			Button.B -> previous.b to current.b
			Button.X -> previous.x to current.x
			Button.Y -> previous.y to current.y
			Button.LB -> previous.left_bumper to current.left_bumper
			Button.RB -> previous.right_bumper to current.right_bumper
			Button.LT -> (previous.left_trigger >= TRIGGER_THRESHOLD) to (current.left_trigger >= TRIGGER_THRESHOLD)
			Button.RT -> (previous.right_trigger >= TRIGGER_THRESHOLD) to (current.right_trigger >= TRIGGER_THRESHOLD)
			Button.UP -> previous.dpad_up to current.dpad_up
			Button.DOWN -> previous.dpad_down to current.dpad_down
			Button.LEFT -> previous.dpad_left to current.dpad_left
			Button.RIGHT -> previous.dpad_right to current.dpad_right
			Button.BACK -> previous.back to current.back
			Button.GUIDE -> previous.guide to current.guide
			Button.START -> previous.start to current.start
			Button.LS -> previous.left_stick_button to current.left_stick_button
			Button.RS -> previous.right_stick_button to current.right_stick_button
		}
}